### ISSUE TITLE
Opam publish fails due to wget retrieval failure (404)

### ISSUE THREAD

**Issue by nik312123** [ocaml-opam/opam-publish#119]
Simply put, I have a library with a module. I can install it locally with `opam install .` in the local version of the GitHub repository without issue and use it just fine.

Following the instructions [here](https://opam.ocaml.org/doc/Packaging.html), I created the corresponding `.opam` file, tested it using `opam install .` and tried using it via `dune`, and tagged and pushed the archive.

However, when I try to run `opam publish`, I get the following error:

```bash
[ERROR] Uncaught exception: OpamDownload.Download_fail(_, "Download command failed: \"/usr/local/bin/wget --content-disposition -t 3 -O
        /private/var/folders/zj/jsz2c_2x4blgtf_b14lfwl3r0000gn/T/opam-29567-5a61c5/archives/790f017e8273931983fa69cb4a5b6951-1.0.tar.gz.part https://github.com/nik312123/[packageName]/archive/1.0.tar.gz -U opam/2.0.7\" exited with code 8
        \"2020-12-18 01:52:47 ERROR 404: Not Found.\"")
```

Here are the versions of various potentially-related packages:

- `wget`: GNU Wget 1.20.3 built on darwin18.7.0.
- `opam`: 2.0.7
- `dune`: 2.7.1
- `git`: git version 2.29.2
- `opam-publish`: 2.0.2
- `gpatch`: 2.7.6

Here is how my `.opam` file generally looks with some redactions as necessary: [example.opam.zip](https://github.com/ocaml-opam/opam-publish/files/5714436/example.opam.zip)

**Comment by rjbou** [comment]
Is your project private ? It seems that the archive is not available (or not publicly). You can test with the same commandline:
```
/usr/local/bin/wget --content-disposition -t 3 -O /<sometmpdir>/downloaded_archive.tar.gz https://github.com/nik312123/[packageName]/archive/1.0.tar.gz"

**Comment by nik312123** [comment]
Ah. It was at the time. It is no longer private. I didn't know that it was required that the repository must be public for the publishing to work. Thank you!

**Comment by nik312123** [comment]
Actually, now the process continues, but another error shows along the way:

```bash
The following will be published:
  - example version 1.0 with opam file at [pathTo/example.opam]
    archive at https://github.com/nik312123/[example]/archive/1.0.tar.gz


Continue ? You will be shown the patch before submitting. [Y/n] y
No Github token found. Should I generate one ?
(your credentials will be required. If you do not have a Github account, you can create one at https://github.com/join) [Y/n] y
The obtained token will be stored in ~/.opam/plugins/opam-publish/ocaml%opam-repository.token.
Your active tokens can be seen and revoked at https://github.com/settings/tokens

Please enter your Github name: nik312123
nik312123 password: 
[ERROR] HTTP Error 404 Not Found
        Headers:
        access-control-allow-origin: *
        access-control-expose-headers: ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Used, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval,
        X-GitHub-Media-Type, Deprecation, Sunset
        content-length: 74
        content-security-policy: default-src 'none'
        content-type: application/json; charset=utf-8
        date: Tue, 22 Dec 2020 14:43:20 GMT
        referrer-policy: origin-when-cross-origin, strict-origin-when-cross-origin
        server: GitHub.com
        status: 404 Not Found
        strict-transport-security: max-age=31536000; includeSubdomains; preload
        vary: Accept-Encoding, Accept, X-Requested-With
        x-content-type-options: nosniff
        x-frame-options: deny
        x-github-media-type: github.v3; format=json
        x-github-request-id: F39C:2A71:43989BE:70A24DA:5FE20608
        x-ratelimit-limit: 60
        x-ratelimit-remaining: 58
        x-ratelimit-reset: 1608651781
        x-ratelimit-used: 2
        x-xss-protection: 1; mode=block

        Body:
        {"message":"Not Found","documentation_url":"https://docs.github.com/rest"}
```

**Comment by rjbou** [comment]
In fact, opam publish prepare the package for publication, without any needed credential (so the archive, repo, etc. need to be reachable), and once the package is ready (opam file updated, linted), then it asks for credential for the branch push and merge request *over* opam repository git repo.
And in your case, the archive is needed to populate `url` section in the opam file, the archive and its md5/sha.

**Comment by nik312123** [comment]
It fails using the form `opam publish https://github.com/nik312123/[example]/archive/1.0.tar.gz .` as well.

**Comment by rjbou** [comment]
It's the same issue than #104 . What is your installed `ocaml-github` version ?

**Comment by nik312123** [comment]
Could you clarify what `ocaml-github` is? I cannot execute `ocaml-github` on the command line.

**Comment by rjbou** [comment]
Yup sorry, I took shortcut. ocaml-github is a dependency to opam-publish. You can have which version is installed with `opam list ocaml-github --installed`.

**Comment by nik312123** [comment]
I got the following output for that command:
```bash
# Packages matching: installed & name-match(ocaml-github)
# No matches found
```

Also, I chose to create a token using https://github.com/settings/tokens and used that instead of putting my username and password. That seemed to work fine. (It for some reason was unable to autogenerate a token through my credentials, though.)

**Comment by nik312123** [comment]
If you meant `opam list github --installed`, then the version is 4.3.2.

**Comment by Aaylor** [comment]
First, sorry for the delay, I did not had the bandwith yet to take a look.

I've took a quick look on how `opam-publish` works and how it interacts with `ocaml-github`; the issue is not totally on `ocaml-github` side, but also on `GitHub` side.

Since last year, the Authorizations API (used to generate tokens) have been deprecated by GitHub (see their [blog post](https://developer.github.com/changes/2019-11-05-deprecated-passwords-and-authorizations-api/#authenticating-using-query-parameters)). And since November 2020, the Authorizations API have been removed completely (see this [blog post](https://github.blog/changelog/2020-11-13-token-authentication-required-for-api-operations/)). Now, it is impossible to auto-generate tokens via their API.

Some work have been done on `ocaml-github` to make the API library uses the token correctly. But the `Token` module (which is used to interact with the Authorization API) has not been deprecated.

Note that some users may still use `opam-publish` with a manually generated token because they already have generated a token using `opam-publish` before the deletion of the Authorization API and are reusing it instead of creating a new one. Note also that the password authentication will not work (as seen by @nik312123), but it really requires a manually generated token.

--

There are multiple that has to be done; by `ocaml-github`:
- I'll first quickly deprecate every API function calling the Authorization API on `ocaml-github` side (to warn used to avoid the usage of these functions that are not working anymore on GitHub side)
- In a major version, remove completely this API and adding documentation on how to authenticate requests

By `opam-publish`:
- remove the usage of the `Token` module (since there are no possibilities to auto-generate tokens anymore), and gives information to user that it is required to manually generate a token (by using this [documentation](https://docs.github.com/en/free-pro-team@latest/github/authenticating-to-github/creating-a-personal-access-token)). This manually generated token can then be given as any `ocaml-github` API function and will be use to authenticate the user.
